---
title: 'SHARP Multiomics Workshop Session 2: Interactions'
author: "Andre Kim"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    df_print: paged
    code_folding: show
  pdf_document: default
editor_options: 
  chunk_output_type: console
---

<!-- ```{css, echo=FALSE} -->
<!-- pre { -->
<!--   max-height: 200px; -->
<!--   overflow-y: auto; -->
<!-- } -->
<!-- ``` -->

```{r setup, include=FALSE, echo=FALSE}
library(tidyverse)
library(MultiAssayExperiment)
library(knitr)
library(ggplot2)
library(reshape2)
library(gap)
library(kableExtra)

options(knitr.table.format = "html")
knitr::opts_chunk$set(echo = TRUE)

# suppress warning messages for final rendering
old.warn <- getOption("warn")
options(warn=-1)
```


# SHARP MULTIOMICS WORKSHOP

## Data descriptions for interactions

As mentioned in the lecture, the HELIX study is a collaborative project across six longitudinal population-based birth cohort studies in six European countries (France, Greece, Lithuania, Norway, Spain, and the United Kingdom). In this example, we will use data from the subcohort of 1,301 mother-child pairs (N = 1,122 with complete data), specifically exposure and genomic data from children 6-11 years old, to investigate GxE interactions between fetal alcohol/smoking exposures and genetic variants, and their influence on future childhood BMI. 

We will consider the following outcome and exposure variables:

* **hs_bmi_c_cat (Outcome)** - Body mass index, dichotomized
  + 0 = Thinness/Normal
  + 1 = Overweight/Obese (WHO reference)
* **lifestyles_e3_alcpreg_yn_None (Interaction exposure)** - Maternal alcohol use during pregnancy
  + 0 = Never
  + 1 = Ever
* **h_mbmi_None** - Maternal pre-pregnancy body mass index
  + kg/m2 (continuous)
* **covariates_e3_sex_None** - Sex
  + 0 = Female
  + 1 = Male
* **h_age_None** - Maternal age
  + years (continuous)
* **h_cohort** - Cohort country
  + 1 = France 
  + 2 = Greece
  + 3 = Lithuania
  + 4 = Norway
  + 5 = Spain 
  + 6 = UK
* **h_edumc_None** - Maternal education
  + 1 = primary school
  + 2 = secondary school
  + 3 = university degree or higher
* **ethn_PC1/2** - principal components (ethnicity)

<br>

### Processing the data

We begin by obtaining the variables of interest from `HELIX.MultiAssayExperiment`, which is an R object that conveniently stores multiple datasets (e.g. exposure, omics) measured on the same individual.  

```{r GxE: Processing the Data, echo=TRUE, cache = T}
# ---- Load RData containing HELIX MultiAssayExperiment object ---- #
load("~/Google Drive/My Drive/SHARP.MultiomicsWorkshop.USC.share/data/HELIX.MultiAssayExperiment.RData")

# ---- specify variable names to extract ---- #
# Outcome (BMI)
outcome.Name <- "hs_bmi_c_cat"
# Exposure (maternal alcohol consumption)
exposure.Name <- "e3_alcpreg_yn_None"
# Covariates
covariate.Names <- c("h_mbmi_None","e3_sex_None","h_age_None","h_cohort","h_edumc_None","ethn_PC1","ethn_PC2")
# Genotypes (m = 1000)
snp.Names <- paste("SNP", 1:1000, sep=".")
variables <- c(covariate.Names, exposure.Name, "h_ethnicity_cauc", snp.Names)

# ---- extract variables (functions from MultiAssayExperiment package ---- #
# 1) select variables but keep in MultiAssayExperiment format
# 2) intersectionColumns selects only individuals with complete data (N = 1,122)
# 3) wideFormat returns as a data.frame
# d <- wideFormat(intersectColumns(helix_ma[variables, ,]), colDataCols=outcome.Name)
# saveRDS(d, "data/d_alcohol.rds")
d <- readRDS("data/d_alcohol.rds")

# Create design matrix
X <- d[,grep("SNP", names(d))]
names(X) <- snp.Names
X <- as.matrix(X)

# Create the outcome variable
Y <- d[,outcome.Name] # outcome
if(outcome.Name=="hs_bmi_c_cat") { Y <- ifelse(as.numeric(Y)>=3, 1, 0)}

# Create the covariate design matrix
U <- d[,c(paste0("covariates_", covariate.Names[1:5]), paste0("proteome.cov_", covariate.Names[6:7]))]
names(U) <- covariate.Names
U[,c("h_cohort","e3_sex_None","h_edumc_None")] <- lapply(U[,c("h_cohort","e3_sex_None","h_edumc_None")], factor)
U[,c("h_mbmi_None", "h_age_None","ethn_PC1","ethn_PC2")] <- lapply(U[,c("h_mbmi_None", "h_age_None","ethn_PC1","ethn_PC2")], as.numeric)
U <- model.matrix(as.formula(paste("~-1+", paste(covariate.Names, collapse="+"))), data=U)
U = U[,-which(colnames(U) == "e3_sex_Nonemale")]

# Other variables for analysis
N <- nrow(d) # number of individuals in the analysis
Q <- ncol(U)  # number of covariates in the matrix U
P <- ncol(X)  # number of SNPs in the matrix X

# Exposure variable
E <- as.numeric(d[,grep(exposure.Name, names(d))])
E <- ifelse(E > 0,1,0) # Dichotomize (smoking)
```

### Descriptive statistics

```{r echo = T, message=F}
library(table1) # table1 package for convenience

dataset <- data.frame(d[,c("primary", "hs_bmi_c_cat", "lifestyles_e3_alcpreg_yn_None",
"covariates_h_mbmi_None", "covariates_e3_sex_None", "covariates_h_age_None",
"covariates_h_cohort", "covariates_h_edumc_None", "proteome.cov_ethn_PC1",
"proteome.cov_ethn_PC2", "proteome.cov_h_ethnicity_cauc")])

# rename columns (for consistency)
names(dataset) <- c("primary", "hs_bmi_c_cat", "e3_alcpreg_yn_None",
"h_mbmi_None", "e3_sex_None", "h_age_None",
"h_cohort", "h_edumc_None", "ethn_PC1",
"ethn_PC2", "h_ethnicity_cauc")

# dichotomize outcome variable
dataset$hs_bmi_c_cat <- ifelse(as.numeric(dataset$hs_bmi_c_cat)>=3, 1, 0)


# code variables properly, apply labels
dataset[, c("h_mbmi_None", "h_age_None", "ethn_PC1", "ethn_PC2")] <- lapply(dataset[, c("h_mbmi_None", "h_age_None", "ethn_PC1", "ethn_PC2")], as.numeric)

dataset[, c("e3_alcpreg_yn_None", "e3_sex_None", "hs_bmi_c_cat", "h_edumc_None", "h_cohort")] <- lapply(dataset[, c("e3_alcpreg_yn_None", "e3_sex_None", "hs_bmi_c_cat", "h_edumc_None", "h_cohort")], factor)

# labels for table1 .. 
dataset$h_edumc_None <- factor(dataset$h_edumc_None, label = c("Primary school", "Secondary school", "University degree or higher"))
dataset$hs_bmi_c_cat <- factor(dataset$hs_bmi_c_cat, label = c("Thin/Normal", "Overweight/Obese"))

label(dataset$e3_alcpreg_yn_None) <- "Maternal alcohol during pregnancy"
label(dataset$h_age_None) <- "Maternal age"
label(dataset$e3_sex_None) <- "Child sex"
label(dataset$h_mbmi_None) <- "Maternal pre-pregnancy BMI"
label(dataset$h_edumc_None) <- "Maternal education"


# run table1 function
table1(~ e3_alcpreg_yn_None +
         h_age_None + 
         e3_sex_None +
         h_edumc_None +
         h_mbmi_None | hs_bmi_c_cat, data=dataset)

```

<br>

### Univariate association tests

```{r, echo = T, results='asis', message = F}

# ---- model hs_bmi_c_cat ~ exposures ---- #
univariate_vars <- c("e3_alcpreg_yn_None", "h_age_None", "e3_sex_None", "h_edumc_None", "h_mbmi_None")

univariate.results <- t(sapply(univariate_vars, FUN=function(p) {  # using index p facilitate write
  x <- dataset[,p]
  reg <- glm(dataset$hs_bmi_c_cat ~ as.numeric(x), family=binomial)    # perform logistic regression
  s.reg <- summary(reg)                 # get the summary for the regression
  c.reg <- s.reg$coef[2,]             # select the coefficients for the exposure
  return(c.reg)                         # to avoid potential memory issues only return coefficients if small number of exposures
}, simplify=T))
univariate.results <- data.frame(univariate_vars,univariate.results)
names(univariate.results) <- c("Variable","Estimate", "SD","Z.Statistic", "P-value")
univariate.results$`P-value` <-  format(univariate.results$`P-value`, scientific=T)

kable(univariate.results) %>% 
  kable_styling()
```

<br>

### Genomic data {.tabset}

Genotype data includes `r P` single nucleotide polymorphisms (SNPs)

#### Plot of genetic ancestry (PCA)
```{r Genome: PC plot, echo=TRUE}
plot(d$proteome.cov_ethn_PC1, d$proteome.cov_ethn_PC2, pch=16, col=ifelse(d$proteome.cov_h_ethnicity_cauc=="yes", 1, 2),
     xlab="Component 1", ylab="Component 2")
legend(x="topleft", legend=c("Caucasian", "Other"), col=c(1,2), pch=16)
```

#### Correlation matrix for local genomic region
```{r Genome: cor.plot, echo=TRUE}
cormat <- round(cor(X[,1:(P/5)], use="complete.obs"), 2)
cormat[lower.tri(cormat)]<- NA
melted_cormat <- melt(cormat)
ggplot(data = melted_cormat, aes(Var2, Var1, fill = value))+
  geom_tile(color = "white")+
  scale_fill_gradient2(low = "blue", high = "red", mid = "white",
                       midpoint = 0, limit = c(-1,1), space = "Lab",
                       name="Pearson\nCorrelation") +
  theme_minimal()+
  theme(axis.text.x = element_blank(), axis.text.y = element_blank())+
  labs(y= "SNPs", x = "SNPs")+
  coord_fixed()

```

<br>

## GxE overview

Question: is there effect heterogeneity for the association of maternal alcohol consumption and childhood BMI by germline genetics?

<br>

### Interaction model

A simple example - fit traditional logistic regression with an interaction term in the form $$logit(Pr D=1|G,E) = \alpha + \beta_EE + \beta_GG + \beta_{GxE}G*E$$

Testing the hypothesis $H0:\beta_{GxE} = 0$ is equivalent to testing whether the ratio $\frac{OR_{GxE}}{OR_G*OR_E} = 1$ e.g. departure from multiplicative effects

In this example, let's start with `SNP.1`: 
```{r}

SNP.1 <- X[,1]

model <- glm(hs_bmi_c_cat ~ e3_alcpreg_yn_None * SNP.1 + h_age_None + e3_sex_None + h_mbmi_None + h_cohort + h_edumc_None, data = dataset, family = 'binomial')

summary(model)

```

The term `e3_alcpreg_yn_None1:SNP.1` has p.value = 0.48. Hence, we cannot reject the null hypothesis and thus cannot conclude there is statistically significant evidence of interactions on a multiplicative scale. 

### Genome-wide interaction scans

To test for interactions genome-wide, we fit the same traditional logistic regression models for every SNP. In this example where m = `r P`, we can use a simple loop:

```{r GxE: Univariate model, echo=TRUE}

  int.results <- t(sapply(1:ncol(X), FUN=function(p) {  # using index p faciliate write
    g <- X[,p]
    reg <- glm(Y~g*E+U, family=binomial)    # perform logistic regression
    s.reg <- summary(reg)                 # get the summary for the regression
    c.reg <- as.numeric(s.reg$coef["g:E",])             # select the coefficients
    # write.table(t(c(paste("SNP", p, sep="."), c.reg)), file="IntResultsGenome.txt", append=ifelse(p==1, F, T), quote=F, sep="\t", col.names=ifelse(p==1, T, F), row.names=F)
    return(c.reg)                         # to avoid potential memory issues only return coefficients if small number of exposures
  }, simplify=T))
  int.results <- data.frame(paste("E:SNP", 1:ncol(X), sep="."),int.results)
  names(int.results) <- c("E:SNP.Name","Estimate", "SD","Z.statistic", "P.value")

```

#### GxE Univariate results: {.tabset}
##### GxE Univariate Summary Table:
```{r GxE: Univariate table}
kable(int.results[int.results$P.value<0.05,], digits=3, align="c", row.names=FALSE, col.names=c("E:SNP","Estimate", "SD","Z statistic", "P value"))
```

##### GxE Univariate Manhattan Plot:
```{r GxE: Univariate MH plot}
neglog.pvalues <- -log10(int.results$P.value)
plot(1:nrow(int.results), neglog.pvalues,
     pch=16, xaxt="n", ylim=c(0, max(neglog.pvalues, 3)),
     ylab="-log(p-value)", xlab="SNPs")
abline(h=-log10(0.05/nrow(int.results)), lty=2, lwd=2, col=2)
```

##### GxE Univariate QQ-Plot:
```{r GxE: QQ-plot}
pvalues <- int.results$P.value
r <- gcontrol2(pvalues, pch=16)
lambda <- round(r$lambda,3)
text(x=1, y=5, labels=bquote(lambda == .(lambda)), cex=2)
```

<br>
